R/sampler-adaptive.R

```diff
@@ -0,0 +1,310 @@
+  ## to merge these over multiple chains?
+  finalise <- function(state, model, rng) {
+    as.list(internal)
+  }
+
+  mcstate_sampler("Adaptive Metropolis-Hastings",
+                  initialise,
+                  step,
+                  finalise)
+}
+
+
+calc_scaling_increment <- function(d, acceptance_target, log_scaling_update) {
+  if (log_scaling_update) {
+    A <- - qnorm(acceptance_target / 2)
+    
+    scaling_increment <- 
+      (1 - 1 / d) * (sqrt(2 * pi) * exp(A ^ 2 / 2)) / (2 * A) + 
+      1 / (d * acceptance_target * (1 - acceptance_target))
+  } else {
+    scaling_increment <- 1 / 100
+  }
+  
+  scaling_increment
+}
+
+
+calc_n_start <- function(acceptance_target) {
+  5 / (acceptance_target * (1 - acceptance_target))
+}
+
+
+qp <- function(x) {
+  outer(x, x)
+}
+
+
+calc_proposal_vcv <- function(scaling, vcv, weight, initial_vcv,
+                              initial_vcv_weight) {
+  d <- dim(vcv)[1]
+  
+  weighted_vcv <-
+    ((weight - 1) * vcv + (initial_vcv_weight + d + 1) * initial_vcv) /
+    (weight + initial_vcv_weight + d + 1)
+  
+  2.38 ^ 2 / d * scaling ^ 2 * weighted_vcv
+}
+
+
+check_replacement <- function(iteration, forget_rate, forget_end) {
+  is_forget_step <- floor(forget_rate * iteration) >
+    floor(forget_rate * (iteration - 1))
+  is_before_forget_end <- iteration <= forget_end
+  
+  is_forget_step & is_before_forget_end
+}
+
+
+update_scaling <- function(scaling, iteration, accept_prob, scaling_increment,
+                           min_scaling, acceptance_target, pre_diminish, 
+                           n_start, log_scaling_update) {
+  scaling_change <- scaling_increment * (accept_prob - acceptance_target) /
+    sqrt(n_start + max(0, iteration - pre_diminish))
+  
+  if (log_scaling_update) {
+    scaling <- pmax(min_scaling, scaling * exp(scaling_change))
+  } else {
+    scaling <- pmax(min_scaling, scaling + scaling_change)
+  }
+  
+}
+
+
+update_autocorrelation <- function(pars, weight, autocorrelation, pars_remove) {
+  if (!is.null(pars_remove)) {
+    if (weight > 2) {
+      autocorrelation <-
+        autocorrelation + 1 / (weight - 1) * (qp(pars) - qp(pars_remove))
+    } else {
+      autocorrelation <- autocorrelation + qp(pars) - qp(pars_remove)
+    }
+  } else {
+    if (weight > 2) {
+      autocorrelation <-
+        (1 - 1 / (weight - 1)) * autocorrelation + 1 / (weight - 1) * qp(pars)
+    } else {
+      autocorrelation <- autocorrelation + qp(pars)
+    }
+  }
+  
+  autocorrelation
+}
+
+
+update_mean <- function(pars, weight, mean, pars_remove) {
+  if (!is.null(pars_remove)) {
+    mean <- mean + 1 / weight * (pars - pars_remove)
+  } else {
+    mean <- (1 - 1 / weight) * mean + 1 / weight * pars
+  }
+  
+  mean
+}
+
+
+update_vcv <- function(mean, autocorrelation, weight) {
+  if (weight > 1) {
+    vcv <- autocorrelation - weight / (weight - 1) * qp(mean)
+  } else {
+    vcv <- 0 * autocorrelation
```

Comment:
If `weight == 1` we only have one parameter set so there is no empirical variance/covariance!

Reply:
We initialise with `weight = 0`, so we hit this bit of code in the end of the first iteration.

You're right that if `forget_rate = 1` it would mean that we would always have one parameter set, it is an extreme choice though! Essentially it would mean we always have no empirical covariance component so the shaping part of the algorithm is entirely non-adaptive. The scaling part would still be adaptive however so it could be useful in case you want to use a fixed VCV but you adaptively scale it?